pkg/user/external_server.go

```diff
@@ -0,0 +1,263 @@
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						sendResponse(rw, http.StatusOK, nil)
+					}
+				}
+			case "POST /user/password":
+				var r UpdatePasswordRequest
+				err := readRequestBody(req, &r)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+					http.Error(rw, err.Error(), http.StatusBadRequest)
+				} else {
+					nu, err := um.UpdatePassword(r.UserID, r.NewPassword)
+					if err != nil {
+						l.WithError(err).Error("[ExternalUserManagerServer] User password update failure")
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						sendResponse(rw, http.StatusOK, nu)
+					}
+				}
+			case "POST /user/api-key":
+				var r ResetAPIKeyRequest
+				err := readRequestBody(req, &r)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+					http.Error(rw, err.Error(), http.StatusBadRequest)
+				} else {
+					nu, err := um.ResetAPIKey(r.UserID)
+					if err != nil {
+						l.WithError(err).Error("[ExternalUserManagerServer] User API key reset failure")
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						sendResponse(rw, http.StatusOK, nu)
+					}
+				}
+			case "DELETE /user":
+				var r DeleteRequest
+				err := readRequestBody(req, &r)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+					http.Error(rw, err.Error(), http.StatusBadRequest)
+				} else {
+					err := um.Delete(r.UserID)
+					if err != nil {
+						l.WithError(err).Error("[ExternalUserManagerServer] User delete failure")
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						sendResponse(rw, http.StatusOK, nil)
+					}
+				}
+			case "GET /user/list":
+				var r ListRequest
+				err := readRequestBody(req, &r)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+					http.Error(rw, err.Error(), http.StatusBadRequest)
+				} else {
+					us, err := um.List(r.Selector)
+					if err != nil {
+						l.WithError(err).Error("[ExternalUserManagerServer] User list failure")
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						sendResponse(rw, http.StatusOK, us)
+					}
+				}
+			case "POST /user/authenticate":
+				var ar AuthenticateRequest
+				err := readRequestBody(req, &ar)
+				if err != nil {
+					l.WithError(err).Error("[ExternalUserManagerServer] Bad request")
+					http.Error(rw, err.Error(), http.StatusBadRequest)
+				} else {
+					u, at, err := um.Authenticate(ar.Username, ar.Password)
+					if err != nil {
+						l.WithError(err).Error("[ExternalUserManagerServer] Authentication failure")
+						http.Error(rw, err.Error(), http.StatusInternalServerError)
+					} else {
+						if u == nil {
+							http.Error(rw, "Authentication failed", http.StatusBadRequest)
```

Comment:
this is probably the opposite of what Go style guidelines suggest :D

Reply:
Yeah, I'm terrible at nonlocal exits in error handling (I blame doing too much functional programming!) Fixing that...

Reply:
I sometimes do that as well but then vscode is showing warnings coming from go linter. Maybe you could also install it for your editor? https://github.com/golang/lint or just "golint". It also shows quite a lot of other useful warnings